### ISSUE TITLE
Bitwarden Firefox extension leaks user unique internal UUID to websites

### ISSUE THREAD

**Issue by ruihildt** [bitwarden/clients#11875]
### Steps To Reproduce

1. Go to 'passkeys.io'
2. Open the browser developer tools
3. In the Inspector tab, search for `moz-extension`
4. The highlighted element contains the bitwarden extension unique internal UUID through the `fido2-page-script.js` script loading

### Expected Result

Bitwarden extension Internal UUID shouldn't be leaked to random websites visited by users.

Internal UUID is unique per install, which means any page integrating passkey can use this ID to uniquely fingerprint users.

### Actual Result

The Bitwarden extension UUID is visible to any website integrating passkey and can be used as a unique fingerprint.

### Screenshots or Videos

![image](https://github.com/user-attachments/assets/3f995c2e-220a-4ee4-90ad-e86b997cd411)

### Operating System

Linux

### Web Browser

Firefox

### Build Version

2024.10.1

### Issue Tracking Info

- [X] I understand that work is tracked outside of Github. A PR will be linked to this issue should one be opened to address it, but Bitwarden doesn't use fields like "assigned", "milestone", or "project" to track progress.

**Comment by ruihildt** [comment]
I have no idea why in this context the code can't be loaded as a content script (which would **not** be leaking the internal UUID), but if the webextension can't do what they need, a bug should be opened with Firefox.

**Comment by Thorin-Oakenpants** [comment]
> a bug should be opened with Firefox

it's been known for years: see bugzillas [1372288,1405971,1717671,1717672](https://bugzilla.mozilla.org/buglist.cgi?bug_id=1372288,1405971,1717671,1717672)

Ideally it should never be leaked, or even a per-session UUID as suggested upstream. I'm not an extension dev, and I don't use bitwarden or know it's UX/user-flow .. but if passkeys is the actual cause, is it not feasible to have this disabled on all sites (hardcoded, no option to globally enable), and the user must add a per site exception to allow it

**Comment by cagonzalezcs** [comment]
Chiming in here @justindbaur, migration of the Firefox and Safari browser extensions to MV3 will help with this issue though it won't completely resolve the problem either.

That UUID is the Bitwarden web extension ID, which can be used to [fingerprint user activity](https://extensionworkshop.com/documentation/develop/build-a-secure-extension/) in certain ways... 

Manifest v2 specifically has issues with leveraging the WebAuthn API to use facilitate the passkeys feature in a "non-isolated" content script context. Mv3 introduces the "Execution World" option to the `browser.scripting` API. 

https://developer.chrome.com/docs/extensions/reference/api/scripting#type-ExecutionWorld

The inline menu is a bit of a different story though. Anything that shows UI elements will always need to load that content within `iframe` windows to allow the page to be isolated from the top level frame of the website.

**Comment by cagonzalezcs** [comment]
@randshell

Yes, the new `browser.scripting` API in Manifest V3, used to inject scripts into a website, resolves most of the issues with the passkeys feature from a content script perspective. In manifest v3, you can avoid DOM injection entirely when injecting the passkeys page-script content. Manifest v2 will not facilitate that at all. 

Also, your suggested workaround is definitely a good way to temporarily resolve the issue. You can also use any Chromium-variant browser to when using Bitwarden. Browsers such as Brave and Opera should be using manifest v3. The only exception to that is Microsoft Edge, which seems to still use manifest v2.

**Comment by kushal256** [comment]
Would love to see this fixed, passkeys are gaining adoption these days and they aren't working in Bitwarden now

**Comment by randshell** [comment]
I've found another vector, I think. When you submit a form, there is a popup for adding / updating the Bitwarden entry, or unlocking the vault if that's the case. This seems to work on both Firefox and Chromium. Following is PoC on Chromium.

HTML:
<img width="1593" height="115" alt="Image" src="https://github.com/user-attachments/assets/3efcd923-796e-4aed-936d-f9242956bcf6" />

Payload:
```js
document.getElementById("bit-notification-bar-iframe").src 
```

Devtools JS result:

<img width="667" height="60" alt="Image" src="https://github.com/user-attachments/assets/f2de60c1-970d-48db-9f4b-6ce5b3572ad5" />

**Comment by randshell** [comment]
@cagonzalezcs are there any updates on this by any chance? As mentioned the other day, the leak happens also when disabling the passkeys, unfortunately. I've actually checked this one more time, and it seems that the idea isn't entirely new, and @felschr  thought about it too in this comment from years ago: https://github.com/mullvad/mullvad-browser/issues/74#issuecomment-1806442479. Quoting it here:

> [...] my guess is that it would only be fingerprintable when it render some UI elements on the website, e.g. the bar on the top that asks whether the user wants to save/update the password they just used to register/login somewhere. If that's the case, that would be less of a concern since these websites are more trusted.

It turns out that with some JS you can auto-submit any form, triggering the popup on any arbitrary and potentially malicious website. I've wrote a simple PoC for this, that, in a way, could be done in a more transparent way to the user. The remediation I've found would be the use of a shadow DOM or, as per the MDN, using an iframe with data URL or `srcdoc` attribute.

I'm planning to write a blog post about this, hopefully in the next few days or week. I assume it shouldn't be a problem since this issue is posted publicly and, nonetheless, the privacy concern around leaking extension's UUIDs has been known since years.

**Comment by cagonzalezcs** [comment]
@randshell 

It has been just under a year since I've worked on the Bitwarden extension, so I can't really give any guidance regarding progress on resolving this concern definitively. 

That said, when specifically speaking about the passkeys feature leaking the extension's UUID, you'd only be experiencing that behavior in cases where the web extension is running on manifest v2. The [browser extension handles injection of the page script that overrides the webauthn API through the chrome.scripting API](https://github.com/bitwarden/clients/blob/00dc1702317568276517e7f1afeaef4f6f07974a/apps/browser/src/autofill/fido2/background/fido2.background.ts#L189) which allows for injection of scripts directly into the MAIN world context of a website.

Since manifest v2 does not have access to that API, you are forced into adding a `script` DOM element to the website, exposing the UUID through a `src` reference to the script. 

Honestly, beyond the semantics of how that script is inserted, the real issue I see lies in having to override the `navigator.credentials` API. Browser extensions need to have support for facilitating behavior of passkeys through the browser extensions's `browser`/`chrome` API namespace for this feature to be more secure. 

Now that said, the extension exposes the UUID in other ways as you alluded to here:

> It turns out that with some JS you can auto-submit any form, triggering the popup on any arbitrary and potentially malicious website. I've wrote a simple PoC for this, that, in a way, could be done in a more transparent way to the user. The remediation I've found would be the use of a shadow DOM or, as per the MDN, using an iframe with data URL or srcdoc attribute.

From my understanding of your comment, you're speaking about presentation of the notification bar that appears after submission of new or updated credentials is made for a website form. 

Yeah, I agree with you on how to address the concern, that's part of the reason why we went through the effort of architecting the inline menu UI to render within closed ShadowDOM elements. I imagine the team plans to implement something similar for the notification bar in the future, but I can't speak to that. 

I'd consider reaching out to @jprusik or @coroiu regarding your concerns. They might be able to provide context on progress in resolving this kind of issue.

**Comment by jprusik** [comment]
A brief update on this: we're currently investigating a few paths to remediation (ideally, moving Firefox builds to manifest v3). I'll keep this thread updated as we make progress.

**Comment by jprusik** [comment]
We have some updates to address this in the short-term at https://github.com/bitwarden/clients/pull/16903

longer-term, we're aspirationally looking to migrating the Firefox client to Manifest v3, but there are some other blocking concerns there, at the moment.

**Comment by n0099** [comment]
https://old.reddit.com/r/firefox/comments/9xnz6p/webextensions_in_firefox_suffer_from_a_privacy/
https://bugzilla.mozilla.org/show_bug.cgi?id=1405971

**Comment by randshell** [comment]
It looks like another change was pushed, which only leaks the UUID during passkey authentication, and that requires user input. While the underlying issue still exists on Firefox-based browsers, at least now the UUID is revealed to websites where you log-in with the passkey.

pre-auth:

<img width="1597" height="896" alt="Image" src="https://github.com/user-attachments/assets/1c333aaa-6231-4727-a68d-3cd881bddcd7" />


during auth:

<img width="1597" height="896" alt="Image" src="https://github.com/user-attachments/assets/f15d870e-6df0-4685-8426-c371d04dc194" />

.
.
Personally, I'll still stay clear of the passkey feature for now.
